### PR TITLE
[crypto] Added Method That Finds and Replaces Resigned Version of a Certificate

### LINKED ISSUE #24211: [Core] Add Method to Detect Resigned Version of a Certificate
### Reproduction steps / Feature

This method is needed and can be used to identify resigned version of PAI or DAC certificate in the ecosystem specific implementations. 

This method can be useful when Matter introduces attestation certificate revocation mechanism.

The following three conditions SHOULD be satisfied for the certificate to qualify as a resigned version of a reference certificate:
      - SKID of the candidate and the reference certificate should match.
      - SubjectDN of the candidate and the reference certificate should match.
      - NotBefore of the candidate certificate should be greater or equal than the
        NotBefore of the reference certificate.

### Platform

core (please add to version below)

### Platform Version(s)

1.0.1

### Type

Unit tested

### (Optional) If manually tested please explain why this is only manually tested

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by emargolis**
Fixes #24211 

This method checks for resigned version of the reference certificate in the list and returns it.

The following conditions SHOULD be satisfied for the certificate to qualify as a resigned version of a reference certificate:
  - SKID of the candidate and the reference certificate should match.
  - SubjectDN of the candidate and the reference certificate should match.

There is no specific use case for this method in Matter. However, specific ecosystem implementations may find this method useful.

Some of the potential use cases could be finding resigned version of a PAI or DAC certificate.

Also, this method can be useful when Matter introduces attestation certificate revocation mechanism.

### CODE REVIEW

File: src/crypto/CHIPCryptoPAL.h [hunk 1]
```diff
@@ -1492,6 +1492,37 @@ CHIP_ERROR ExtractSKIDFromX509Cert(const ByteSpan & certificate, MutableByteSpan
...
+CHIP_ERROR ExtractAKIDFromX509Cert(const ByteSpan & certificate, MutableByteSpan & akid);
+
+/**
+ * @brief Checks for resigned version of the certificate in the list and returns it.
+ *
+ * The following three conditions SHOULD be satisfied for the certificate to qualify as
+ * a resigned version of a reference certificate:
+ *   - SKID of the candidate and the reference certificate should match.
+ *   - SubjectDN of the candidate and the reference certificate should match.
+ *   - NotBefore of the candidate certificate should be greater or equal than the
```
Comment (tcarmelveilleux) [inline src/crypto/CHIPCryptoPAL.h:1507]: Not sure we should validate this last condition. The certs in the replacement set can just be made to have valid dates. In both the case where the notbefore is checked and not checked, providing erroneous certs would make the substitution fail. Not accounting for trying to validate the notbefore date would simplify the code.

Comment (emargolis) [reply src/crypto/CHIPCryptoPAL.h]: I agree, for simplicity I removed the time validity checks and the comparison will be based only on SKID and Subject.

File: src/crypto/CHIPCryptoPAL.h [hunk 2]
```diff
@@ -1492,6 +1492,32 @@ CHIP_ERROR ExtractSKIDFromX509Cert(const ByteSpan & certificate, MutableByteSpan
  **/
 CHIP_ERROR ExtractAKIDFromX509Cert(const ByteSpan & certificate, MutableByteSpan & akid);
 
+/**
+ * @brief Extracts the Authority Key Identifier from an X509 Certificate.
+ **/
+CHIP_ERROR ExtractAKIDFromX509Cert(const ByteSpan & certificate, MutableByteSpan & akid);
```
Comment (bzbarsky-apple) [inline src/crypto/CHIPCryptoPAL.h:1498]: This function is already declared 5 lines earlier, no?  Merge issue?

File: src/crypto/CHIPCryptoPAL.h [hunk 3]
```diff
@@ -1492,6 +1492,32 @@ CHIP_ERROR ExtractSKIDFromX509Cert(const ByteSpan & certificate, MutableByteSpan
...
+
+/**
+ * @brief Checks for resigned version of the certificate in the list and returns it.
+ *
+ * The following conditions SHOULD be satisfied for the certificate to qualify as
+ * a resigned version of a reference certificate:
+ *   - SKID of the candidate and the reference certificate should match.
+ *   - SubjectDN of the candidate and the reference certificate should match.
+ *
+ * If no resigned version is found then reference certificat itself is returned.
```
Comment (bzbarsky-apple) [inline src/crypto/CHIPCryptoPAL.h:1508]:
```suggestion
 * If no resigned version is found then reference certificate itself is returned.
```

File: src/crypto/CHIPCryptoPALOpenSSL.cpp [hunk 4]
```diff
@@ -2045,5 +2045,65 @@ CHIP_ERROR ExtractVIDPIDFromX509Cert(const ByteSpan & certificate, AttestationCe
     return err;
 }
 
+CHIP_ERROR ReplaceCertIfResignedCertFound(const ByteSpan & referenceCertificate, const ByteSpan * candidateCertificates,
+                                          size_t candidateCertificatesCount, ByteSpan & outCertificate)
+{
+    CHIP_ERROR err                              = CHIP_NO_ERROR;
+    X509 * x509ReferenceCertificate             = nullptr;
+    X509 * x509CandidateCertificate             = nullptr;
+    const unsigned char * pReferenceCertificate = referenceCertificate.data();
```
Comment (bzbarsky-apple) [inline src/crypto/CHIPCryptoPALOpenSSL.cpp:2054]:
```suggestion
    const uint8_t * pReferenceCertificate = referenceCertificate.data();
```

File: src/crypto/CHIPCryptoPALOpenSSL.cpp [hunk 5]
```diff
@@ -2045,5 +2045,65 @@ CHIP_ERROR ExtractVIDPIDFromX509Cert(const ByteSpan & certificate, AttestationCe
...
+    x509ReferenceCertificate = d2i_X509(nullptr, &pReferenceCertificate, static_cast<long>(referenceCertificate.size()));
+    VerifyOrExit(x509ReferenceCertificate != nullptr, err = CHIP_ERROR_NO_MEMORY);
+
+    referenceSubject = X509_get_subject_name(x509ReferenceCertificate);
+    VerifyOrExit(referenceSubject != nullptr, err = CHIP_ERROR_INTERNAL);
+
+    for (size_t i = 0; i < candidateCertificatesCount; i++)
+    {
+        const ByteSpan candidateCertificate         = candidateCertificates[i];
+        const unsigned char * pCandidateCertificate = candidateCertificate.data();
```
Comment (bzbarsky-apple) [inline src/crypto/CHIPCryptoPALOpenSSL.cpp:2079]:
```suggestion
        const uint8_t * pCandidateCertificate = candidateCertificate.data();
```

File: src/crypto/CHIPCryptoPALTinyCrypt.cpp [hunk 6]
```diff
@@ -1691,5 +1691,85 @@ CHIP_ERROR ExtractVIDPIDFromX509Cert(const ByteSpan & certificate, AttestationCe
...
+
+        if (referenceSKID.data_equal(candidateSKID) && x509_string_cmp(&referenceSubject, &candidateSubject) == 0)
+        {
+            outCertificate = candidateCertificate;
+            return CHIP_NO_ERROR;
+        }
+    }
+
+#else
+    (void) referenceCertificate;
```
Comment (bzbarsky-apple) [inline src/crypto/CHIPCryptoPALTinyCrypt.cpp:1765]: Should this case really silently violate the API contract for this function?

File: src/crypto/CHIPCryptoPALmbedTLS.cpp [hunk 7]
```diff
@@ -1836,5 +1836,85 @@ CHIP_ERROR ExtractVIDPIDFromX509Cert(const ByteSpan & certificate, AttestationCe
...
+
+        if (referenceSKID.data_equal(candidateSKID) && x509_string_cmp(&referenceSubject, &candidateSubject) == 0)
+        {
+            outCertificate = candidateCertificate;
+            return CHIP_NO_ERROR;
+        }
+    }
+
+#else
+    (void) referenceCertificate;
```
Comment (bzbarsky-apple) [inline src/crypto/CHIPCryptoPALmbedTLS.cpp:1910]: Again, this seems pretty suspicious.

File: src/platform/silabs/EFR32/CHIPCryptoPALPsaEfr32.cpp [hunk 8]
```diff
@@ -1875,5 +1875,85 @@ CHIP_ERROR ExtractVIDPIDFromX509Cert(const ByteSpan & certificate, AttestationCe
...
+
+        if (referenceSKID.data_equal(candidateSKID) && x509_string_cmp(&referenceSubject, &candidateSubject) == 0)
+        {
+            outCertificate = candidateCertificate;
+            return CHIP_NO_ERROR;
+        }
+    }
+
+#else
+    (void) referenceCertificate;
```
Comment (bzbarsky-apple) [inline src/platform/silabs/EFR32/CHIPCryptoPALPsaEfr32.cpp:1949]: And here.

File: src/crypto/CHIPCryptoPAL.h (hunk 3 above)
Comment (emargolis) [reply src/crypto/CHIPCryptoPAL.h:1508]: done

File: src/crypto/CHIPCryptoPALOpenSSL.cpp (hunk 4 above)
Comment (emargolis) [reply src/crypto/CHIPCryptoPALOpenSSL.cpp:2054]: done

File: src/crypto/CHIPCryptoPALOpenSSL.cpp (hunk 5 above)
Comment (emargolis) [reply src/crypto/CHIPCryptoPALOpenSSL.cpp:2079]: done

File: src/crypto/CHIPCryptoPALTinyCrypt.cpp (hunk 6 above)
Comment (emargolis) [reply src/crypto/CHIPCryptoPALTinyCrypt.cpp:1765]: This implementation matches all other method implementations in this file. I would suggest to review and fix it as part of other PR.

File: src/crypto/CHIPCryptoPALmbedTLS.cpp (hunk 7 above)
Comment (emargolis) [reply src/crypto/CHIPCryptoPALmbedTLS.cpp:1910]: This implementation matches all other method implementations in this file. I would suggest to review and fix it as part of other PR.

File: src/platform/silabs/EFR32/CHIPCryptoPALPsaEfr32.cpp (hunk 8 above)
Comment (emargolis) [reply src/platform/silabs/EFR32/CHIPCryptoPALPsaEfr32.cpp:1949]: This implementation matches all other method implementations in this file. I would suggest to review and fix it as part of other PR.

**Review by bzbarsky-apple — APPROVED** [review]

File: src/crypto/CHIPCryptoPAL.h [hunk 9]
```diff
@@ -1492,6 +1492,27 @@ CHIP_ERROR ExtractSKIDFromX509Cert(const ByteSpan & certificate, MutableByteSpan
...
+ *   - SKID of the candidate and the reference certificate should match.
+ *   - SubjectDN of the candidate and the reference certificate should match.
+ *
+ * If no resigned version is found then reference certificate itself is returned.
+ *
+ *  @param referenceCertificate       A DER certificate.
+ *  @param candidateCertificates      A pointer to the list of DER Certificates, which should be searched
+ *                                    for the resigned version of `referenceCertificate`.
+ *  @param candidateCertificatesCount Number of certificates in the `candidateCertificates` list.
+ *  @param outCertificate             A reference to the certificate or it's resigned version if found.
```
Comment (tcarmelveilleux) [inline src/crypto/CHIPCryptoPAL.h:1509]: Please note that it points to either referenceCertificate or one of candidateCertificates, and it's not a copy.

**Review by tcarmelveilleux — APPROVED** [review]

File: src/crypto/CHIPCryptoPAL.h (hunk 9 above)
Comment (emargolis) [reply src/crypto/CHIPCryptoPAL.h:1509]: added clarification

**Pull request merged**